### ISSUE TITLE
Template and dynamic jobs not available when updating resource_paths via Project().state.update()

### ISSUE THREAD

**Issue by s4b7r** [pyiron/pyiron_base#750]
Like we have worked out in #589 I use 

```py
from pyiron import Project
pr = Project(path='my_proj')
pr.state.update({
    'sql_file': './pyiron.db',
    'sql_type': 'SQLite',
})
```

to use a database file that's in my folder for development. I now try to extend the resource_paths setting to make use of template / dynamic job classes that are also local to my development folder by also updating `'resource_paths': str(Path('.') / 'pyiron' / 'resources')`. `pr.state.settings.configuration` reflects those changes. However, template and dynamic jobs in ./pyiron/resources are still not available.

I assume this is because dynamic job classes are only loaded in pyiron_base.generic.dynamic._get_template_classes() and that is only called once, when pyiron_base.job.jobtype loads the generic.dynamic module and fills JOB_DYN_DICT. Can you confirm this assumption?

Is there any way around this behaviour?

**Comment by pmrv** [comment]
> I assume this is because dynamic job classes are only loaded in pyiron_base.generic.dynamic._get_template_classes() and that is only called once, when pyiron_base.job.jobtype loads the generic.dynamic module and fills JOB_DYN_DICT. Can you confirm this assumption?

Hi, good work on diagnosing the issue, I think that's what's happening here! The solution should be to redefine the `JOB_DYN_DICT` when the resource paths change.  I'll patch this early next week.  As a hack inbetween this will probably work
```python
from pyiron_base.generic.dynamic import _get_template_classes
from pyiron import Project
pr = Project(...)
pr.create.job._job_dyn_dict = _get_template_classes()
```
You will need to do that for every (python) instance of a project you create.

**Comment by niklassiemer** [comment]
Actually, it should be enough to do
```
from pyiron_base.generic.dynamic import _get_template_classes, JOB_DYN_DICT

JOB_DYN_DICT=_get_template_classes()
```
Once for the python shell after you changed the resorces.

**Comment by pmrv** [comment]
This will not update the job wrapper though, right?

**Comment by niklassiemer** [comment]
I would need to check again. However, as far as I recall the wrapper uses the `JOB_*_DICT`s to populate the dir method. I might be wrong and mixing this up with my wished behavior, though.

**Comment by pmrv** [comment]
It does use it, but because of how python variables work, your example won't update it correctly.

Basically python variables are pointers to the actual object.  E.g. when `python_base.generic.dynamic` is first imported, it runs a line like
```python
JOB_DYN_DICT=_get_template_classes()
```
which does actually *two* things.  It will create the new dictionary somewhere in memory and then make `JOB_DYN_DICT` point to that location.  This is always how variable assignment works, evaluate the right hand side, then make the left hand side point to that result.  This means that "overwriting" a variable will never touch the object it has previously pointed to.
In our case here the `JobType` imports the `JOB_DYN_DICT` like this
```python
from pyiron_base.generic.dynamic import JOB_DYN_DICT, class_constructor
```
which you can imagine to work like
```python
import pyiron_base.generic.dynamic 
JOB_DYN_DICT = pyiron_base.generic.dynamic.JOB_DYN_DICT
class_constructor = pyiron_base.generic.dynamic.class_constructor
del pyiron_base.generic.dynamic 
```
We then have two variables ("pointers", `pyiron_base.generic.dynamic.JOB_DYN_DICT` and `pyiron_base.generic.job.jobtype.JOB_DYN_DICT`) that point to the `dict` object created by the first call to `_get_template_classes()`, but because setting a variable simply updates where it's pointing to, importing either and overwriting it, will most affect the other variable at all.  The situation would be different if `JobType` used the qualified name, then you could re-assign `pyiron_base.generic.dynamic.JOB_DYN_DICT` and `JobType` would see this.

Anyway, sorry for the text wall.

TL;DR: Python's variables are a bit weird, it looks completely by-value, but is actually by-reference most of the time.

**Comment by s4b7r** [comment]
Thanks for your replies. The workaround seems to work. I have to fix some follow up errors on my side of the code to do a final check.

Thanks also for the nice write-up on Python's behavior. I knew about Python's interesting take on by-referenc'ing almost everything, but never thought about it in context of how names are actually imported into the local namespace.

**Comment by s4b7r** [comment]
> ```python
> from pyiron_base.generic.dynamic import _get_template_classes
> from pyiron import Project
> pr = Project(...)
> pr.create.job._job_dyn_dict = _get_template_classes()
> ```

Can confirm now, this works.

**Comment by s4b7r** [comment]
Sorry, I was a bit fast closing this issue. If you consider that behavior a bug, you can reopen this issue. For the time being I am fine with the workaround.

**Comment by pmrv** [comment]
Yeah, this should not happen, but I won't get around to the fix this week.  Let's keep it open.

**Comment by samwaseda** [comment]
`_get_template_classes` does not exist anymore, but instead (for now) `create_job_class` should be used (cf. [this example](https://github.com/pyiron/pyiron_base/blob/c99d951a730df91d17d949f56437e8d51789e57e/pyiron_base/jobs/flex/executablecontainer.py#L16)). I copy and paste the example:

```python
import os

def write_input(input_dict, working_directory="."):
    with open(os.path.join(working_directory, "input_file"), "w") as f:
        f.write(str(input_dict["energy"]))


def collect_output(working_directory="."):
    with open(os.path.join(working_directory, "output_file"), "r") as f:
        return {"energy": float(f.readline())}


from pyiron_base import Project
pr = Project("test")
pr.create_job_class(
    class_name="CatJob",
    write_input_funct=write_input,
    collect_output_funct=collect_output,
    default_input_dict={"energy": 1.0},
    executable_str="cat input_file > output_file",
)
job = pr.create.job.CatJob(job_name="job_test")
job.input["energy"] = 2.0
job.run()
```
